src/logging/drain.rs

```diff
@@ -0,0 +1,58 @@
+use anyhow::Result;
+use slog::{OwnedKVList, Record};
+
+use super::event;
+
+/// A logging drain designed to integrate with [`slog::Drain`]
+///
+/// The drain can be easily added to a regular [`slog::Logger`]:
+///
+/// ```rust
+/// use kubewarden_policy_sdk::logging;
+/// use slog::{Drain, Logger, o};
+///
+/// let drain = logging::KubewardenDrain::new().fuse();
+/// let log = Logger::root(drain, o!("logger_key1" => "logger_value1"));
+/// ```
+///
+/// The drain behaves differently based on the target architecture used at
+/// build time.
+///
+/// The `wasm32` target architecture will cause the drain to use the [`wapc_guest::host_call`]
+/// helper to propagate the log event from the Wasm guest to the native host
+/// environment. This is the default behaviour for Kubewarden policies at
+/// execution time.
+///
+/// Building for a non `wasm32` architecture will cause the drain to print the log
+/// entries on the standard output.
+/// This is useful for running test of policies via a regular `cargo test`.
+pub struct KubewardenDrain {}
+
+impl KubewardenDrain {
+    #![allow(clippy::new_without_default)]
+    pub fn new() -> KubewardenDrain {
+        KubewardenDrain {}
+    }
+}
+
+impl slog::Drain for KubewardenDrain {
+    type Ok = ();
+    type Err = anyhow::Error;
+
+    #[cfg(not(target_arch = "wasm32"))]
+    fn log(&self, rinfo: &Record, logger_values: &OwnedKVList) -> Result<()> {
+        let event = event::new(rinfo, logger_values).unwrap();
+        println!("GOT: {}", serde_json::to_string(&event)?);
+
+        Ok(())
+    }
+
+    #[cfg(target_arch = "wasm32")]
+    fn log(&self, rinfo: &Record, logger_values: &OwnedKVList) -> Result<()> {
+        let event = event::new(rinfo, logger_values).unwrap();
+        let msg = serde_json::to_vec(&event).unwrap();
+        let _res = wapc_guest::host_call("kubewarden", "tracing", "log", &msg);
```

Comment:
In light of an error, shouldn't we propagate it?

Reply:
This code runs inside of the wasm code, if something goes wrong we have no way to propagate the error to the host. Printing to the stdout would not work, because we are not using Wasi.

Do you think we should do one more attempt at sending a generic error?

Reply:
Yeah, I think we don't have much of a choice here. No, I would not try to do more attempts.

The reason why I I think propagating is important is to stop the flow of the guest module if we cannot log. We don't know if we are trying to log something important, so I think we should propagate up the chain.

Reply:
Sorry, I got it now. I've implemented it

Reply:
I'm now propagating the result